spring-integration-core/src/main/java/org/springframework/integration/util/MessagingAnnotationUtils.java

```diff
@@ -69,9 +70,31 @@
 		return null;
 	}
 
-	public static boolean hasValue(Object value) {
-		return value != null && (!(value instanceof String) || (StringUtils.hasText((String) value))) // NOSONAR
-				&& (!value.getClass().isArray() || ((Object[]) value).length > 0);
+	/**
+	 * Determines if the value of a named attribute from an annotation instance contains an actual value.
+	 * <p>It is considered to NOT contain a value if it is null, an empty string, an empty array, or a {@link Poller}
+	 * whose 'value' field is set to {@link ValueConstants#DEFAULT_NONE}.
+	 *
+	 * @param annotationValue the value of the annotation attribute
+	 * @return whether {@code annotationValue} is set to a non-empty value
+	 */
+	public static boolean hasValue(Object annotationValue) {
+		if (annotationValue == null) {
+			return false;
+		}
+		// Empty array
+		if (ObjectUtils.isArray(annotationValue) && ObjectUtils.isEmpty(annotationValue)) {
+			return false;
+		}
+		// Empty string
+		if ((annotationValue instanceof String) && ObjectUtils.isEmpty(annotationValue)) {
+			return false;
+		}
+		// Poller with special 'none' value
+		if ((annotationValue instanceof Poller) && ValueConstants.DEFAULT_NONE.equals(((Poller) annotationValue).value())) {
```

Comment:
This is not OK. 
Better really to have an utility method as generics as possible.
I mean I believe there is a way to get a value from the annotation without casting to some specific type.
For example see the same `@Reactive` use-case:
```
Reactive reactive() default @Reactive(ValueConstants.DEFAULT_NONE);
```
It is fully similar to what we have so far with the `@Poller`.


Also: if we do this then we don't need that extra check `if (poller != null && !ValueConstants.DEFAULT_NONE.equals(poller.value())) {` in the `AbstractMethodAnnotationPostProcessor.java`

Reply:
Yep, I meant to add a comment on this controversial section. The alternative approach I was going to talk about would be to consider any annotation whose "value" field was set to `ValueConstants.DEFAULT_NONE` to be considered "empty". This would take the `Poller` or `Reactive` type out of the equation. This would also handle `@Reactive` annotation extension properly. 

> For example see the same @Reactive use-case: 
> It is fully similar to what we have so far with the @Poller.

AFAICT we do not handle `@Reactive` override in annotation hierarchy properly (you will see it is not in the EnableIntegrationTests). We would need to add the special handling for `@Reactive` as well. 

I will amend the proposal w/ the above approach and we can see how that looks.